### PR TITLE
Add MethodsV2 unit testing

### PR DESCRIPTION
**Pull request by tgreenx**
## Purpose

This PR adds MethodsV2 unit testing, based on test zone specification from https://github.com/zonemaster/zonemaster/pull/1254. All external methods are covered.

## Context

Test zone specification: https://github.com/zonemaster/zonemaster/pull/1254

## Changes

- Add unit testing for MethodsV2
- Minor fix in `Zonemaster::Engine::TestMethodsV2`

## How to test this PR

Unit tests are updated and should pass.

### CODE REVIEW

**Comment by matsduf** [comment]
I disagree that only testing "Get zone NS names and IP addresses" gives a full coverage. The extraction of parent IPs is not tested. Delegation and zone do not need to match. Delegation can be with or without glue or parts of glue.

**Comment by matsduf** [comment]
We could limit the data in the `t` file to be

* Get parent NS IP addresses
* Get delegation NS names and IP addresses
* Get zone NS names and IP addresses

From "Get delegation NS names and IP addresses" the following two can be extracted:
* Get delegation NS names
* Get delegation NS IP addresses

From "Get zone NS names and IP addresses" the following two can be extracted:
* Get zone NS names
* Get zone NS IP addresses

We should verify that MethodsV2 do that extraction correctly, but maybe not for every scenario.

File: t/TestUtil.pm [hunk 1]
```diff
@@ -70,22 +106,256 @@ an array of DS expressions for "undelegated" DS
...
 =back
 
-The array of mandatory message tags or the array of forbidden message tags, but not both, could be
-undefined. If the mandatory message tag array is undefined, then it will be generated to contain
-all message tags not included in the forbidden message tag array. The same mechanism is used if the
-forbidden message tag array is undefined.
+The arrays of mandatory message tags and forbidden message tags, but not both, could be undefined.
+If the mandatory message tag array is undefined, then it will be generated to contain all message
+tags not included in the forbidden message tag array. The same mechanism is used if the forbidden
+message tag array is undefined.
```
Comment (marc-vanderwal) [inline t/TestUtil.pm:112]: Is this change supposed to be part of this pull request? I seem to have reviewed a similar passage before and suggesting an improvement. After sleeping on it, I think I’ve come up with something better still.

```suggestion
It is possible to pass C<undef> instead of the mandatory message tag or forbidden message tag arrayref.
In that case, it is equivalent to an arrayref containing C<@$aref_alltags> minus the tags in the other array.
In other words, passing C<undef> as the forbidden message tag arrayref means that any tag not explicitly
allowed must not be emitted, while passing C<undef> as the mandatory message tag arrayref means
that any tag not explicitly forbidden must be emitted.

It is an error if both of the aforementioned arrayrefs are simultaneously undefined or simultaneously empty.
```

File: t/methodsv2.t [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+#
+
+my %subtests = (
+    'GOOD-1' => [
+        1,
+        q(child.parent.good-1.methodsv2.xa),
+        [ qw( 127.40.3.21 fda1:b2:c3:0:127:40:3:21 127.40.3.22 fda1:b2:c3:0:127:40:3:22 ) ],
+        [ qw( ns1.child.parent.good-1.methodsv2.xa/127.40.4.21 ns1.child.parent.good-1.methodsv2.xa/fda1:b2:c3:0:127:40:4:21 ns2.child.parent.good-1.methodsv2.xa/127.40.4.22 ns2.child.parent.good-1.methodsv2.xa/fda1:b2:c3:0:127:40:4:22 ) ],
+        [ qw( ns1.child.parent.good-1.methodsv2.xa/127.40.4.21 ns1.child.parent.good-1.methodsv2.xa/fda1:b2:c3:0:127:40:4:21 ns2.child.parent.good-1.methodsv2.xa/127.40.4.22 ns2.child.parent.good-1.methodsv2.xa/fda1:b2:c3:0:127:40:4:22 ) ],
+        []
```
Comment (marc-vanderwal) [inline t/methodsv2.t:43]: It would be much nicer to read if this data is on multiple lines, like this:
```suggestion
        [ qw( 127.40.3.21 fda1:b2:c3:0:127:40:3:21
              127.40.3.22 fda1:b2:c3:0:127:40:3:22 ) ],
        [ qw( ns1.child.parent.good-1.methodsv2.xa/127.40.4.21
              ns1.child.parent.good-1.methodsv2.xa/fda1:b2:c3:0:127:40:4:21
              ns2.child.parent.good-1.methodsv2.xa/127.40.4.22
              ns2.child.parent.good-1.methodsv2.xa/fda1:b2:c3:0:127:40:4:22 ) ],
        [ qw( ns1.child.parent.good-1.methodsv2.xa/127.40.4.21
              ns1.child.parent.good-1.methodsv2.xa/fda1:b2:c3:0:127:40:4:21
              ns2.child.parent.good-1.methodsv2.xa/127.40.4.22
              ns2.child.parent.good-1.methodsv2.xa/fda1:b2:c3:0:127:40:4:22 ) ],
        []
```

**Review by marc-vanderwal — COMMENTED** [review]
Looks fine, though I have some minor suggestions.

**Comment by matsduf** [comment]
I have realized that I have done less optimal selection of IP addresses for the scenarios. The selection principle is also different for the test scenarios for test cases. I am about to renumber, both in https://github.com/zonemaster/zonemaster/pull/1254 and as a PR to this PR.

**Comment by matsduf** [comment]
For scenarios for test cases it is possible to test and trouble shoot with the help of `zonemaster-cli`, but that is not possible for the MethodsV2 scenarios. Would it be possible to at a "verbose" mode for the `t` file in which only one scenario is run and verbose information is outputted, e.g. expected name servers and found name servers, respectively?

**Comment by matsduf** [comment]
I have created https://github.com/tgreenx/zonemaster-engine/pull/2 to update this PR with the new IP addresses from an update of https://github.com/zonemaster/zonemaster/pull/1254. A new data file is also created.

**Comment by matsduf** [comment]
I tested by changing the correct to an incorrect. I tested scenario `GOOD-1`.

In  `EXPECTED_PARENT_IP` I changed "127.40.1.41" to "127.40.2.41". In `EXPECTED_DEL_NS` I changed "ns1.child.parent.good-1.methodsv2.xa/127.40.1.51" to "ns1.childs.parent.good-1.methodsv2.xa/127.40.1.51".

I got the following output:

```
        #   Failed test 'Nameserver IP '127.40.2.41' is present'
        #   at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 270.
        # Nameserver IP '127.40.2.41' should have been present, but wasn't
        # Looks like you failed 1 test of 6.

    #   Failed test 'get_parent_ns_ips'
    #   at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 278.
```
I expects it to say that the look up data gives "127.40.1.41" that was not found in the expected data.

```
        #   Failed test 'Nameserver 'ns1.childs.parent.good-1.methodsv2.xa/127.40.1.51' is present'
        #   at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 291.
        # Nameserver 'ns1.childs.parent.good-1.methodsv2.xa/127.40.1.51' should have been present, but wasn't
        # Looks like you failed 1 test of 6.

    #   Failed test 'get_del_ns_names_and_ips'
    #   at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 299.
```

Same here. "ns1.child.parent.good-1.methodsv2.xa/127.40.1.51" is what we get from the look-up, but we do not get it.

**Comment by matsduf** [comment]
I removed one IP address from `EXPECTED_PARENT_IP` and got the following error message:

```
        #   Failed test at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 273.
        # Number of nameserver IP addresses in both arrays does not match
        # Looks like you failed 1 test of 5.

    #   Failed test 'get_parent_ns_ips'
    #   at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 278.
    # Looks like you failed 1 test of 7.
t/methodsv2.t .. 4/? 
#   Failed test 'GOOD-1'
```

It is not very helpful. Are there too many in `EXPECTED_PARENT_IP` or too few? Which is the problematic IP address?

Then I also changed one address in `EXPECTED_PARENT_IP` but that information is not provided.

```
        #   Failed test at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 273.
        # Number of nameserver IP addresses in both arrays does not match
        # Looks like you failed 1 test of 5.

    #   Failed test 'get_parent_ns_ips'
    #   at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 278.
    # Looks like you failed 1 test of 7.

#   Failed test 'GOOD-1'
```

**Comment by matsduf** [comment]
I removed one IP address from `EXPECTED_PARENT_IP` and got the following error message:

```
        #   Failed test at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 273.
        # Number of nameserver IP addresses in both arrays does not match
        # Looks like you failed 1 test of 5.

    #   Failed test 'get_parent_ns_ips'
    #   at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 278.
    # Looks like you failed 1 test of 7.
t/methodsv2.t .. 4/? 
#   Failed test 'GOOD-1'
```

It is not very helpful. Are there too many in `EXPECTED_PARENT_IP` or too few? Which is the problematic IP address?

Then I also changed one address in `EXPECTED_PARENT_IP` but that information is not provided.

```
        #   Failed test at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 273.
        # Number of nameserver IP addresses in both arrays does not match
        # Looks like you failed 1 test of 5.

    #   Failed test 'get_parent_ns_ips'
    #   at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 278.
    # Looks like you failed 1 test of 7.

#   Failed test 'GOOD-1'
```

**Comment by matsduf** [comment]
Scenario GOOD-6 fails. See #1351.

**Comment by matsduf** [comment]
IB-NOT-IN-ZONE-1 fails and also triggers "Use of uninitialized value".

```
t/methodsv2.t .. 11/? 
        #   Failed test 'Nameserver 'ns1.child.parent.ib-not-in-zone-1.methodsv2.xa' is present'
        #   at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 291.
        # Nameserver 'ns1.child.parent.ib-not-in-zone-1.methodsv2.xa' should have been present, but wasn't

        #   Failed test 'Nameserver 'ns2.child.parent.ib-not-in-zone-1.methodsv2.xa' is present'
        #   at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 291.
        # Nameserver 'ns2.child.parent.ib-not-in-zone-1.methodsv2.xa' should have been present, but wasn't

        #   Failed test at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 294.
        # Number of nameservers in both arrays does not match
        # Looks like you failed 3 tests of 4.

    #   Failed test 'get_zone_ns_names_and_ips'
    #   at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 299.
Use of uninitialized value $expected_ip in concatenation (.) or string at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 339.

        #   Failed test 'Nameserver IP '' is present'
        #   at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 339.
Use of uninitialized value $expected_ip in concatenation (.) or string at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 339.
        # Nameserver IP '' should have been present, but wasn't

        #   Failed test at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 342.
        # Number of nameserver IPs in both arrays does not match
        # Looks like you failed 2 tests of 3.

    #   Failed test 'get_zone_ns_ips'
    #   at /home/ubuntu/git/zonemaster-engine/t/TestUtil.pm line 347.
    # Looks like you failed 2 tests of 7.
t/methodsv2.t .. 14/? 
#   Failed test 'IB-NOT-IN-ZONE-1'
```

**Comment by matsduf** [comment]
I suggest that the error message is something like:

```
Unexpected: 127.40.3.21 
Unexpected: fda1:b2:c3:0:127:40:3:21
Missing: 127.40.3.22
Missing: fda1:b2:c3:0:127:40:3:22
```

"Unexpected" when not in the expected set, but given by the method. "Missing" when in the expected set, but not given by the method.

**Comment by matsduf** [comment]
The current error message only says what is expected but not found, but not was found and not expected.

**Comment by matsduf** [comment]
> @matsduf I added a fix ([48ff5dc](https://github.com/zonemaster/zonemaster-engine/commit/48ff5dc3d0bbdb9a30b643c4851bf8eca69ae0dc)) to a crash in method `get_zone_ns_names_and_ips()`. It could happen when an out-of-bailiwick nameserver had no IP address (thus being of type `Zonemaster::Engine::DNSName` instead of `Zonemaster::Engine::Nameserver`). Could we have such a test scenario defined in [zonemaster/zonemaster#1254](https://github.com/zonemaster/zonemaster/pull/1254) ?

Your request for a scenario when "an out-of-bailiwick nameserver had no IP address" could be interpreted as either of the following two:

1. There is no glue record in the delegation, but the name can be resolved by lookup.
2. There is no glue record in the delegation and the name cannot be resolved.

For alternative 1 there is already scenario GOOD-2. For alternative 2 there are no scenarios, but I will add such scenarios. "Has no IP address" can be either NODATA or NXDOMAIN and I will create both. If the zone has two NS it could be on one or both. There are many combinations!

File: lib/Zonemaster/Engine/TestMethodsV2.pm [hunk 3]
```diff
@@ -69,7 +69,7 @@ This Method will obtain the IP addresses of the Out-Of-Bailiwick name servers fo
 
 Takes a Zonemaster::Engine::Zone object (C<$zone>) and an arrayref of Zonemaster::Engine::Nameserver objects (C<$ns_names_ref>).
 
-Returns an arrayref of Zonemaster::Engine::Nameserver objects.
+Returns an arrayref of Zonemaster::Engine::Nameserver and/or Zonemaster::Engine::DNSName objects.
```
Comment (marc-vanderwal) [inline lib/Zonemaster/Engine/TestMethodsV2.pm:72]: Two suggestions:
 - Wrap module names in `L<…>` so that they become hyperlinks when viewing the documentation online (e.g. on Metacpan);
 - Make it clearer when the function returns a Zonemaster::Engine::Nameserver object and when it returns a Zonemaster::Engine::DNSName.

```suggestion
Returns an arrayref of L<Zonemaster::Engine::Nameserver> objects for each name server name that was successfully resolved to an IP address, and L<Zonemaster::Engine::DNSName> objects for each name server name that could not be resolved to an IP address.
```

File: lib/Zonemaster/Engine/TestMethodsV2.pm [hunk 4]
```diff
@@ -267,7 +267,7 @@ If the Glue Records include address records for Out-Of-Bailiwick name servers th
 
 Takes a Zonemaster::Engine::Zone object (C<$zone>).
 
-Returns an arrayref of Zonemaster::Engine::Nameserver objects, or `undef` if no parent zone was found.
+Returns an arrayref of Zonemaster::Engine::Nameserver and/or Zonemaster::Engine::DNSName objects, or `undef` if no parent zone was found.
```
Comment (marc-vanderwal) [inline lib/Zonemaster/Engine/TestMethodsV2.pm:270]: Same suggestion as above.

File: lib/Zonemaster/Engine/TestMethodsV2.pm [hunk 5]
```diff
@@ -450,7 +450,7 @@ For the In-Bailiwick name server names obtain the IP addresses from the given zo
...
```
Comment (marc-vanderwal) [inline lib/Zonemaster/Engine/TestMethodsV2.pm:453]: Same suggestion as above.

**Comment by matsduf** [comment]
@tgreenx, there is a textual conflict in `t/TestUtil.pm`. Could you resolve that?

**Review by matsduf — APPROVED** [review]
